### PR TITLE
Training workload for Llama 3.2 1B - DoRA

### PR DESCRIPTION
**Pull request by upantelicTT**
### Ticket
Closes https://github.com/tenstorrent/tt-blacksmith/issues/236

### Model training setting

- **Model:** `Llama 3.2 1B`
- **Strategy:** SFT with DoRA layers on a single chip (N150)
- **Framework:** JAX
- **Dataset:** SST2

### What's changed
Implement DoRA fine-tuning workload for Llama 3.2 1B

### CODE REVIEW

File: blacksmith/experiments/jax/llama_dora/lorax/helpers.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        m = col_norms.reshape(m_shape).astype(param.dtype)
+
+        return DoraWeight(param, a, b, m, alpha=alpha)
+
+    return jax.tree_util.tree_map_with_path(get_param, param_tree, spec, is_leaf=is_leaf)
+
+
+def simple_spec(params, decision_fn=None, tune_vectors=False, is_leaf=None) -> Any:
+    """
+    Create a simple dora spec for a pytree
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/lorax/helpers.py]:
```suggestion
    Create a simple DoRA spec for a pytree.
```

File: blacksmith/experiments/jax/llama_dora/lorax/helpers.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        value = decision_fn(path, arr)
+        return value
+
+    return tree_map_with_path(full_fn, params, is_leaf=is_leaf)
+
+
+def split_trainable_frozen(dora_params, dora_spec) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+    """
+    Split DoRA parameters into trainable and frozen pytrees.
+    Trainable: Only DoRA a,b matrices from DoraWeight objects
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/lorax/helpers.py]: nit: Please add this comment either in doc or in code, duplication seems unnecessary.

File: blacksmith/experiments/jax/llama_dora/lorax/transform.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+
+@quax.register(lax.dot_general_p)
+def handle_dot_lhs_dora(dora: DoraWeight, rhs: jax.Array, *, dimension_numbers: Any, **kwargs: Any) -> Any:
+    """
+    Handle DoRA forward pass: y = (m * normalize(W + B @ A)) @ x
+    """
+    if isinstance(rhs, DoraWeight):
+        rhs = rhs.materialise()
+        warnings.warn("Encountered product of two DoraWeights. Materializing the rhs")
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/lorax/transform.py]: Log should come before the call of the code. If something breaks, you would not get the log.

File: blacksmith/experiments/jax/llama_dora/lorax/transform.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    v_scaled = dora.m * v_normed
+
+    out = op(lhs, v_scaled, dimension_numbers=dimension_numbers)
+    return out.astype(dora.w.dtype)
+
+
+@quax.register(lax.transpose_p)
+def eval_dora_transpose(arg: DoraWeight, *, permutation: Any) -> Any:
+    """
+    Define how a `DoraWeight` behaves under transpose.
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/lorax/transform.py]:
```suggestion
    Define how `DoraWeight` behaves under transpose.
```

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+
+import jax
+import jax.numpy as jnp
+import optax
+import numpy as np
+from transformers import FlaxAutoModelForCausalLM, AutoTokenizer, AutoConfig
+
+from datasets import load_dataset
+import lorax
+from lorax import DORA_FREEZE
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: nit: It seems like an overkill to import constant like this just to use it once.

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+def main() -> None:
+    """Main training function with configurable parameters."""
+
+    config_file_path = os.path.join(os.path.dirname(__file__), "test_llama_fine_tuning_jax.yaml")
+    training_config = generate_config(TrainingConfig, config_file_path)
+
+    cpu_device = jax.devices("cpu")[0]
+    current_device, device_kind = _select_preferred_device()
+
+    print(f"Loading Llama 3.2-1B model... Using device: {device_kind} -> {current_device}")
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: Are you actually using `device_kind`, it feels like you are using cpu?

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    # Initializing model parameters on CPU, since jax.random.normal
+    # is currently not supported on device (https://github.com/tenstorrent/tt-xla/issues/1105).
+    with jax.default_device(cpu_device):
+        model = load_model(training_config.model_name)
+
+    model.params = jax.tree_util.tree_map(lambda x: jax.device_put(x, current_device), model.params)
+
+    wandb_run = setup_wandb(training_config, enable=training_config.model_to_wandb, device=device_kind)
+
+    input_id_batches, attention_mask_batches, label_batches = load_data(training_config)
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: If you print loading of the model, you should print loading of the dataset as well.

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+
+    except Exception as e:
+        print(f"Error during training: {e}")
+        log_to_wandb({"error": str(e), "training_failed": True})
+        raise
+
+    finally:
+        if WANDB_ENABLED and wandb is not None:
+            wandb.finish()
+            print("Finished wandb run")
```
Comment (mmilosevicTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]:
```suggestion
            print("Finished wandb run.")
```

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+from blacksmith.experiments.torch.llama.configs import TrainingConfig
+from blacksmith.tools.cli import generate_config
+import wandb
+
+MODEL_NAME = "Erland/Llama-3.2-1B-JAX"
+DEFAULT_EXPERIMENT_NAME = "Llama-TT-DoRA-Training"
+DEFAULT_RUN_NAME = "llama-3.2-1b-sst2-tt-lorax"
+
+
+WANDB_ENABLED = False
```
[the remark this answers is not in the document]
Comment (pmarkovicTT) [reply blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: same for constants above, make sure to add them to the config (check other experiments for reference)

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+
+def load_model(model_name: str) -> FlaxAutoModelForCausalLM:
+    """Load and configure the Llama model for training."""
+    config = AutoConfig.from_pretrained(model_name)
+    config.use_cache = False
+    config.dtype = jnp.float32
+    return FlaxAutoModelForCausalLM.from_pretrained(model_name, config=config, from_pt=False, dtype=jnp.float32)
+
+
+def load_data(training_config: TrainingConfig) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray]:
```
Comment (pmarkovicTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: Should be part of data loader class, make sure to add data loader as specified above

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+
+    input_id_batches, attention_mask_batches, label_batches = load_data(training_config)
+
+    decision_fn = create_dora_decision_fn(training_config.lora_r)
+    dora_spec = lorax.simple_spec(model.params, decision_fn=decision_fn, tune_vectors=False)
+
+    # Initializing model parameters on CPU, since jax.random.normal
+    # is currently not supported on device (https://github.com/tenstorrent/tt-xla/issues/1105).
+    with jax.default_device(cpu_device):
+        dora_params = lorax.init_dora(model.params, dora_spec, jax.random.PRNGKey(42))
```
[the remark this answers is not in the document]
Comment (pmarkovicTT) [reply blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: is it a `seed`?

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+        train_labels.append(np.array(item["labels"]))
+
+    train_input_ids = create_batches(jnp.array(train_input_ids), training_config.batch_size)
+    train_attention_mask = create_batches(jnp.array(train_attention_mask), training_config.batch_size)
+    train_labels = create_batches(jnp.array(train_labels), training_config.batch_size)
+
+    return train_input_ids, train_attention_mask, train_labels
+
+
+def create_dora_decision_fn(dora_rank: int) -> Any:
```
Comment (pmarkovicTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: It seems to me this should be part of model loading file? Definitely no reason to be inside training script file

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+    finally:
+        if WANDB_ENABLED and wandb is not None:
+            wandb.finish()
+            print("Finished wandb run")
+
+    print("Testing sentiment classification generation....")
+
+
+if __name__ == "__main__":
+    main()
```
Comment (pmarkovicTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: Please use main part/guard to init config, wandb and other configurations, then call` train() ` function (renamed `main` function). Check torch llama fine tuning script.

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+    trainable_params, frozen_params = lorax.split_trainable_frozen(dora_params, dora_spec)
+
+    with jax.default_device(cpu_device):
+        optimizer = optax.adamw(learning_rate=training_config.learning_rate, weight_decay=0.01)
+        trainable_params_cpu = jax.tree_util.tree_map(lambda x: jax.device_put(x, cpu_device), trainable_params)
+        opt_state = optimizer.init(trainable_params_cpu)
+
+    dora_model = lorax.dora(model)
+    loss_fn = create_loss_fn(dora_model)
+    compute_grads_tt = create_compute_grads_fn(loss_fn)
```
Comment (umalesTT) [inline blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: nit: When you change the names of functions, please remove "_tt" from function name as well

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+import numpy as np
+from transformers import FlaxAutoModelForCausalLM, AutoTokenizer, AutoConfig
+
+from datasets import load_dataset
+import lorax
+from lorax import DORA_FREEZE
+
+from blacksmith.datasets.torch.llama.sst_dataset import SSTDataset
+from blacksmith.experiments.torch.llama.configs import TrainingConfig
+from blacksmith.tools.cli import generate_config
```
[the remark this answers is not in the document]
Comment (mmilosevicTT) [reply blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: Well, the link you provided states this:
```
However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:
```
In my view, adding `blacksmith` to each import is not adding any value whatsoever, as all our files are in the same source directory. On top of that, I do not agree that it is easy to interpret and refactor. If you decide to change the name of the top directory, you have to add additional tens of lines of code to your change. Non-sibling modules stay the same when it comes to explicit and non-explicit notation, you are only removing LCA of imports to keep it compact.

Since this is a nit comment, I am fine with doing it both ways, but am still saying that adding repeated instances of a single value in all imports is not a good practice.

[the remark this answers is not in the document]
Comment (pglusacTT) [reply blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: Yes, it does state that, however the example below shows:

> from . import sibling
> from .sibling import example

and we are doing

> from ....datasets.torch.llama.sst_dataset import SSTDataset

which it is not really applicable to that statement.

Renaming the top-level package is rare and handled by a project-wide refactor. IDEs can easily and safely rename your top-level package and make sure everything is correct. File moves are common and absolute imports are safer and better in those scenarios. 

We can't call this a bad practice, it is actually really common and encouraged. 
Read [django](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style#imports) and [pandas](https://pandas.pydata.org/pandas-docs/version/1.2.4/development/code_style.html#imports-aim-for-absolute) guidelines, for example.

I agree that it is not a big deal, but we should have a clear separation when is it okay to use relative imports for our coding guidelines. 

We should take this offline and discuss our coding guidelines further, the sooner we have them the better.

File: blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+MODEL_NAME = "Erland/Llama-3.2-1B-JAX"
+DEFAULT_EXPERIMENT_NAME = "Llama-TT-DoRA-Training"
+DEFAULT_RUN_NAME = "llama-3.2-1b-sst2-tt-lorax"
+
+
+WANDB_ENABLED = False
+
+
+def setup_wandb(training_config: TrainingConfig, enable: bool = False, device: str = "tt") -> Optional[Any]:
+    """Optionally setup wandb for experiment tracking; returns run or None.
```
[the remark this answers is not in the document]
Comment (umalesTT) [reply blacksmith/experiments/jax/llama_dora/test_llama_fine_tuning_jax.py]: Yes, I agree, I inserted Args and Returns into more complex functions.
